<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Nathan Harmston">
<meta name="dcterms.date" content="2024-05-14">

<title>ChIPQC</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="ChIPQC_files/libs/clipboard/clipboard.min.js"></script>
<script src="ChIPQC_files/libs/quarto-html/quarto.js"></script>
<script src="ChIPQC_files/libs/quarto-html/popper.min.js"></script>
<script src="ChIPQC_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ChIPQC_files/libs/quarto-html/anchor.min.js"></script>
<link href="ChIPQC_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ChIPQC_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ChIPQC_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ChIPQC_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ChIPQC_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">ChIPQC</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Nathan Harmston </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 14, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="chipqc" class="level1">
<h1>ChIPQC</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ChIPQC)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'S4Vectors' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'GenomeInfoDb' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'SummarizedExperiment' was built under R version 4.3.2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set 6 workers</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">register</span>(<span class="fu">MulticoreParam</span>(<span class="at">workers =</span> <span class="dv">6</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Read in samples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Load sample data</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">'../meta/chipqc_metadata.csv'</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  SampleID Factor Replicate                                          bamReads
1      IP1  Deaf1         1 /Users/sbinh/Downloads/deaf1/IP_DEAF1_1_final.bam
2      IP2  Deaf1         2 /Users/sbinh/Downloads/deaf1/IP_DEAF1_2_final.bam
  ControlID                                    bamControl
1    Input1 /Users/sbinh/Downloads/deaf1/Input1_final.bam
2    Input2 /Users/sbinh/Downloads/deaf1/Input2_final.bam
                                                     Peaks PeakCaller Tissue
1 /Users/sbinh/Downloads/deaf1/deaf1_rep1_peaks.narrowPeak narrowPeak     NA
2 /Users/sbinh/Downloads/deaf1/deaf1_rep2_peaks.narrowPeak narrowPeak     NA
  Condition
1        NA
2        NA</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Create ChIPQC object</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>chipObj <span class="ot">&lt;-</span> <span class="fu">ChIPQC</span>(samples, <span class="at">annotation=</span><span class="st">"mm10"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>IP1  Deaf1   1 narrowPeak</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>IP2  Deaf1   2 narrowPeak</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Checking chromosomes:</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "chr1"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Compiling annotation...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing metrics for 4 samples...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Executing in parallel - max 8 cores.</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Create ChIPQC report</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ChIPQCreport</span>(chipObj, <span class="at">reportName=</span><span class="st">"ChIP QC report: Deaf1"</span>, <span class="at">reportFolder=</span><span class="st">"ChIPQCreport"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Saving 7 x 5 in image
Saving 7 x 5 in image
Saving 7 x 5 in image
Using Sample as id variables

Saving 7 x 5 in image
Saving 7 x 5 in image</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Europe/London
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] ChIPQC_1.38.0               BiocParallel_1.36.0        
 [3] DiffBind_3.12.0             SummarizedExperiment_1.32.0
 [5] Biobase_2.62.0              MatrixGenerics_1.14.0      
 [7] matrixStats_1.2.0           GenomicRanges_1.54.1       
 [9] GenomeInfoDb_1.38.6         IRanges_2.36.0             
[11] S4Vectors_0.40.2            BiocGenerics_0.48.1        
[13] ggplot2_3.5.0              

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3                       
  [2] rstudioapi_0.15.0                        
  [3] jsonlite_1.8.8                           
  [4] magrittr_2.0.3                           
  [5] GenomicFeatures_1.54.3                   
  [6] farver_2.1.1                             
  [7] rmarkdown_2.25                           
  [8] ragg_1.2.7                               
  [9] BiocIO_1.12.0                            
 [10] zlibbioc_1.48.0                          
 [11] vctrs_0.6.5                              
 [12] memoise_2.0.1                            
 [13] Rsamtools_2.18.0                         
 [14] RCurl_1.98-1.14                          
 [15] SQUAREM_2021.1                           
 [16] mixsqp_0.3-54                            
 [17] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2  
 [18] htmltools_0.5.7                          
 [19] S4Arrays_1.2.0                           
 [20] progress_1.2.3                           
 [21] curl_5.2.0                               
 [22] truncnorm_1.0-9                          
 [23] SparseArray_1.2.4                        
 [24] KernSmooth_2.23-22                       
 [25] htmlwidgets_1.6.4                        
 [26] plyr_1.8.9                               
 [27] TxDb.Hsapiens.UCSC.hg18.knownGene_3.2.2  
 [28] cachem_1.0.8                             
 [29] GenomicAlignments_1.38.2                 
 [30] lifecycle_1.0.4                          
 [31] pkgconfig_2.0.3                          
 [32] Matrix_1.6-5                             
 [33] R6_2.5.1                                 
 [34] fastmap_1.1.1                            
 [35] GenomeInfoDbData_1.2.11                  
 [36] digest_0.6.35                            
 [37] systemPipeR_2.8.0                        
 [38] numDeriv_2016.8-1.1                      
 [39] colorspace_2.1-0                         
 [40] chipseq_1.52.0                           
 [41] ShortRead_1.60.0                         
 [42] AnnotationDbi_1.64.1                     
 [43] irlba_2.3.5.1                            
 [44] textshaping_0.3.7                        
 [45] RSQLite_2.3.5                            
 [46] hwriter_1.3.2.1                          
 [47] labeling_0.4.3                           
 [48] invgamma_1.1                             
 [49] filelock_1.0.3                           
 [50] fansi_1.0.6                              
 [51] httr_1.4.7                               
 [52] abind_1.4-5                              
 [53] compiler_4.3.1                           
 [54] bit64_4.0.5                              
 [55] withr_3.0.0                              
 [56] DBI_1.2.2                                
 [57] TxDb.Mmusculus.UCSC.mm9.knownGene_3.2.2  
 [58] gplots_3.1.3.1                           
 [59] biomaRt_2.58.2                           
 [60] MASS_7.3-60.0.1                          
 [61] rappdirs_0.3.3                           
 [62] DelayedArray_0.28.0                      
 [63] rjson_0.2.21                             
 [64] GreyListChIP_1.34.0                      
 [65] gtools_3.9.5                             
 [66] caTools_1.18.2                           
 [67] tools_4.3.1                              
 [68] glue_1.7.0                               
 [69] restfulr_0.0.15                          
 [70] grid_4.3.1                               
 [71] reshape2_1.4.4                           
 [72] generics_0.1.3                           
 [73] gtable_0.3.4                             
 [74] BSgenome_1.70.2                          
 [75] hms_1.1.3                                
 [76] xml2_1.3.6                               
 [77] utf8_1.2.4                               
 [78] XVector_0.42.0                           
 [79] ggrepel_0.9.5                            
 [80] pillar_1.9.0                             
 [81] stringr_1.5.1                            
 [82] emdbook_1.3.13                           
 [83] limma_3.58.1                             
 [84] dplyr_1.1.4                              
 [85] BiocFileCache_2.10.1                     
 [86] lattice_0.22-5                           
 [87] rtracklayer_1.62.0                       
 [88] bit_4.0.5                                
 [89] deldir_2.0-2                             
 [90] tidyselect_1.2.1                         
 [91] locfit_1.5-9.8                           
 [92] Biostrings_2.70.2                        
 [93] amap_0.8-19                              
 [94] knitr_1.45                               
 [95] TxDb.Mmusculus.UCSC.mm10.knownGene_3.10.0
 [96] xfun_0.42                                
 [97] statmod_1.5.0                            
 [98] stringi_1.8.3                            
 [99] yaml_2.3.8                               
[100] evaluate_0.23                            
[101] codetools_0.2-19                         
[102] interp_1.1-6                             
[103] bbmle_1.0.25.1                           
[104] TxDb.Rnorvegicus.UCSC.rn4.ensGene_3.2.2  
[105] tibble_3.2.1                             
[106] cli_3.6.2                                
[107] systemfonts_1.0.5                        
[108] munsell_0.5.1                            
[109] Rcpp_1.0.12                              
[110] dbplyr_2.4.0                             
[111] coda_0.19-4.1                            
[112] png_0.1-8                                
[113] bdsmatrix_1.3-6                          
[114] XML_3.99-0.16.1                          
[115] parallel_4.3.1                           
[116] blob_1.2.4                               
[117] prettyunits_1.2.0                        
[118] latticeExtra_0.6-30                      
[119] jpeg_0.1-10                              
[120] bitops_1.0-7                             
[121] mvtnorm_1.2-4                            
[122] apeglm_1.24.0                            
[123] scales_1.3.0                             
[124] Nozzle.R1_1.1-1.1                        
[125] crayon_1.5.2                             
[126] TxDb.Dmelanogaster.UCSC.dm3.ensGene_3.2.2
[127] rlang_1.1.3                              
[128] ashr_2.2-63                              
[129] KEGGREST_1.42.0                          
[130] TxDb.Celegans.UCSC.ce6.ensGene_3.2.2     </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>