---
title: "Annotate deaf1 ChIP-seq peaks"
author: Nathan Harmston
date: '`r format(Sys.time(), "%d %B %Y")`'
format:
  html: 
    keep-md: true
    embed-resources: true
    df-print: kable
    toc: true
    toc-depth: 3
    code-fold: true
    number-sections: true
    smooth-scroll: true
    code-tools: true
    code-line-numbers: true
  gfm: 
    df-print: kable
    toc: true 
    toc-depth: 3
    number-sections: true
---


# Annotation of deaf1 ChIP-seq data

```{r}
suppressPackageStartupMessages({
    # Load libraries
    library(dplyr)
    library(ggplot2)
    library(Matrix)
    library(ChIPseeker)
    library(TxDb.Mmusculus.UCSC.mm10.knownGene)
    library(EnsDb.Mmusculus.v79)
    library(AnnotationDbi)
    library(org.Mm.eg.db)
})  
```


```{r}
library(biomaRt)
library(GenomicRanges)
library(Biostrings)
library(GenomicFeatures)
library(org.Mm.eg.db)

mm.gtf.db <- makeTxDbFromGFF("~/Downloads/Mus_musculus.GRCm38.102.chr.gtf", format="gtf" )

#ensembl.genes = genes(mm.gtf.db)
ensembl.tx = transcripts(mm.gtf.db)
# Use version 102 (which fits our gtf file)

mouse = useEnsembl(biomart="ENSEMBL_MART_ENSEMBL", 
                   dataset="mmusculus_gene_ensembl", 
                  version = 102)

bm.annotations = getBM(attributes=c("ensembl_gene_id", "entrezgene_id", 
                                    "gene_biotype", "external_gene_name",
                                   "ensembl_transcript_id"),
                       mart=mouse, 
                       filters="ensembl_transcript_id", 
                       values=ensembl.tx$tx_name, uniqueRows=TRUE
                      )

ensembl.tx$gene_biotype = bm.annotations$gene_biotype[match(ensembl.tx$tx_name, bm.annotations$ensembl_transcript_id) ]
ensembl.tx$entrezgene_id = bm.annotations$entrezgene_id[match(ensembl.tx$tx_name, bm.annotations$ensembl_transcript_id) ]
ensembl.tx$external_gene_name = bm.annotations$external_gene_name[match(ensembl.tx$tx_name, bm.annotations$ensembl_transcript_id) ]
ensembl.tx$ensembl_transcript_id =  bm.annotations$ensembl_transcript_id[match(ensembl.tx$tx_name, bm.annotations$ensembl_transcript_id) ]
ensembl.tx$ensembl_gene_id <- bm.annotations$ensembl_gene_id[
    match(ensembl.tx$tx_name, bm.annotations$ensembl_transcript_id)
]

head(ensembl.tx)

# save this as RDS so we don't have to keep running it
saveRDS(bm.annotations, file = "output/bm_annotations_tx.RDS")
saveRDS(ensembl.tx, file = "output/ensembl_tx.RDS")
```


```{r}
library(rtracklayer)
library(BSgenome.Mmusculus.UCSC.mm10)

deaf1.idr = read.delim("../peaks/deaf1-idr", header=FALSE)
deaf1.filtered.idr = deaf1.idr[deaf1.idr[,5]>=540,] 
deaf1.filtered.idr = deaf1.filtered.idr[!deaf1.filtered.idr[,1] %in% c("chr4_GL456216_random", "chr4_JH584295_random"), ] 
 
deaf1.gr = GRanges(deaf1.filtered.idr[,1], IRanges(deaf1.filtered.idr[,2], deaf1.filtered.idr[,3]), idr= deaf1.filtered.idr[,5],
                 signalValue= deaf1.filtered.idr[,7])
names(deaf1.gr) = paste("peak", 1:length(deaf1.gr), sep=":")
```



```{r}
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
```

```{r}
# Prepare the promoter regions
promoter <- getPromoters(TxDb = txdb, upstream = 1000, downstream=1000)
tagMatrixList <- getTagMatrix(deaf1.gr, windows = promoter)
```

```{r}
#| label: peak-count
plotAvgProf(tagMatrixList, xlim = c(-1000, 1000),  conf = 0.95,resample = 500, facet = "row") + labs(title = "deaf1 binding sites")
```

```{r}
# Retrieve annotations
peakAnnoList <- annotatePeak(deaf1.gr, TxDb=txdb, 
                       tssRegion=c(-1000, 1000), verbose=TRUE)
```

  

```{r}
peakAnnoList
```



```{r}
#| label: feature-distribution
#| fig-height: 4
#| fig-width: 10

plotAnnoBar(peakAnnoList)
```


```{r}
#| label: dist-to-tss

plotDistToTSS(peakAnnoList, 
              title="Distribution of transcription factor-binding loci \n relative to TSS")
```


## How many unique genes are bound by 608 peaks? 

```{r}
deaf1_annot <- data.frame(peakAnnoList)
head(deaf1_annot)
```

Distribution of binding annotations: 
```{r}
deaf1_annot  %>% 
    group_by(annotation) %>% 
    summarize(n = n())  %>% 
    arrange(desc(n))  %>% 
    slice_head(n = 5)
```

First, we add the ensembl gene ID to the transcript ID: 

```{r}
deaf1_annot <- deaf1_annot %>% 
    tidyr::separate(transcriptId, 
             into = c("ens_transcript", sep = "."))
```

```{r}
deaf1_annot$ens_genes <- ensembl.tx$ensembl_gene_id[
    match(deaf1_annot$ens_transcript, ensembl.tx$ensembl_transcript_id)]
    
deaf1_annot$gene_symbol <- ensembl.tx$external_gene_name[
    match(deaf1_annot$ens_transcript, ensembl.tx$ensembl_transcript_id)]
```

```{r}
deaf1_annot$name = names(deaf1.gr)

```


```{r}
# Check whether any genes are unmapped
sum(is.na(deaf1_annot$ens_genes))
```

0 genes unmapped; all genes are mapped. 

```{r}
write.csv(deaf1_annot, file =  "output/deaf1_annot.csv")
write.csv(deaf1.idr, file =  "output/deaf1_idr.csv", quote=FALSE, row.names=F)
```


How many unique genes? 

```{r}
length(unique(deaf1_annot$ens_genes))
```

# Summary


# sessionInfo()

```{r}
sessionInfo()
```


